node/src/components/block_validator.rs

```diff
@@ -59,114 +65,116 @@ pub(crate) struct BlockValidationState<T> {
-                    .collect();
 
                 // TODO: Clean this up to use `or_insert_with_key` once
                 // https://github.com/rust-lang/rust/issues/71024 is stabilized.
                 match self.validation_states.entry(block) {
                     Entry::Occupied(mut entry) => {
-                        // The entry already exists. We register ourselves as someone interested in
-                        // the ultimate validation result.
-                        entry.get_mut().responders.push(responder);
+                        // The entry already exists.
+                        if entry.get().missing_deploys.is_empty() {
+                            // Block has already been validated successfully, early return to
+                            // caller.
+                            effects.extend(responder.respond((true, entry.key().clone())).ignore());
+                        } else {
+                            // We register ourselves as someone interested in the ultimate
+                            // validation result.
+                            entry.get_mut().responders.push(responder);
+                        }
                     }
                     Entry::Vacant(entry) => {
                         // Our entry is vacant - create an entry to track the state.
                         let missing_deploys: HashSet<DeployHash> =
                             entry.key().deploys().iter().map(|hash| **hash).collect();
 
+                        let in_flight = &mut self.in_flight;
+                        let chainspec = self.chainspec.clone();
+                        let fetch_effects: Effects<Event<T, I>> = block_deploys
+                            .iter()
+                            .flat_map(|deploy_hash| {
+                                let chainspec = chainspec.clone();
+                                // For every request, increase the number of in-flight...
+                                in_flight.inc(deploy_hash);
+
+                                // ...then request it.
+                                let dh_found = *deploy_hash;
+                                let dh_not_found = *deploy_hash;
+                                effect_builder
+                                    .fetch_deploy(*deploy_hash, sender.clone())
+                                    .map_or_else(
+                                        move |result: FetchResult<Deploy>| match result {
+                                            FetchResult::FromStorage(deploy)
+                                            | FetchResult::FromPeer(deploy, _) => {
+                                                if deploy.header().is_valid(
+                                                    &chainspec.genesis.deploy_config,
+                                                    block_timestamp,
+                                                ) {
+                                                    Event::DeployFound(dh_found)
```

Comment:
The level of nesting here is pretty grim - I'd be inclined to try and define this closure in the `flat_map` outside of here or something like that.

Reply:
Done.